src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp

```diff
@@ -128,11 +128,17 @@ class MacroAssembler: public Assembler {
     a.lea(this, r);
   }
 
+  // Whether materializing the given address for a LDR/STR requires an
+  // additional lea instruction.
+  static bool legitimize_address_requires_lea(const Address &a, int size) {
+    return !Address::offset_ok_for_immed(a.offset(), exact_log2(size));
+  }
+
   /* Sometimes we get misaligned loads and stores, usually from Unsafe
      accesses, and these can exceed the offset range. */
   Address legitimize_address(const Address &a, int size, Register scratch) {
     if (a.getMode() == Address::base_plus_offset) {
-      if (! Address::offset_ok_for_immed(a.offset(), exact_log2(size))) {
+      if (legitimize_address_requires_lea(a, size)) {
```

Comment:
It is a little strange that `legitimize_address_requires_lea` is only the second condition and not 
```
  return a.getMode() == Address::base_plus_offset && !Address::offset_ok_for_immed(a.offset(), exact_log2(size));
```

And have the check in `legitimize_address` simply be `if (legitimize_address_requires_lea(a, size))`

I guess we never end up calling `legitimize_address_requires_lea` with a literal address, where it would assert in `a.offset()`.  But requiring the Address parameter of legitimize_address_requires_lea to be in a specific mode as a precondition seems weird to me.

Reply:
Thanks Axel, I fully agree, done (commit cf4f3b30).